PDEP-10: Add pyarrow as a required dependency (#52711)
pandas-dev/pandas: web/pandas/pdeps/0010-required-pyarrow-dependency.md

integration into many areas of pandas. With our roadmap noting that pandas strives for better Apache Arrow
interoperability [^1] and many projects [^2], within or beyond the Python ecosystem, adopting or interacting with
the Arrow format, making PyArrow a required dependency provides an additional signal of confidence in the Arrow
ecosystem to pandas users.

Additionally, requiring PyArrow would simplify the related development within pandas and potentially improve NumPy
functionality that would be better suited by PyArrow including:

- Avoiding runtime checking if PyArrow is available to perform PyArrow object inference during constructor or indexing operations
  - Currently, there are 17 `import_optional_dependency("pyarrow")` checks throughout the pandas code base

- Removing unnecessary functionality:
  - fastparquet engine in ``read_parquet``
  - potentially simplifying the ``read_csv`` logic (needs more investigation)

- Avoiding NumPy object data types more by default for analogous types that have native PyArrow support such as:
  - decimal
  - binary
  - nested types (like lists, dicts, ...)
  - strings

> Out of this group, strings offer the most advantages for users. They use significantly less memory and are faster:

Reviewer: Haven't kept up with this, but how are the plans to add the new numpy string dtype (xref #47884 ) going to affect the rationale here?

I would assume performance of the numpy string dtype would be on par with the pyarrow one.

Reviewer 2: The linked comment said that numpy strings are available "within a year or so". 
This does not seem to be dissimilar to a pandas 3.0 release date now proposed here?

Reviewer 3: > this is still years' away

I hope it doesn't take that long!

> (Nathan - we discussed timelines before, but I didn't write them down so have forgotten them, apologies)

The earliest pandas could officially support the dtype I'm working on is after the release of Numpy 2.0 - currently scheduled for January 2024. This assumes the new dtype API is available for downstream use in Numpy 2.0 without needing to set an environment variable. I'm hoping to start shipping experimental support in pandas behind the environment variable after Numpy 1.25 is released this summer, as that version of Numpy will hopefully have a version of the new dtype API that is usable for pandas' needs. The version in Numpy 1.24 is broken and is missing a lot of features we've added since that release, unfortunately.

Reviewer 3: > I can't remember the perf comparison - @ngoldbaum do you want to comment here?

Creating e.g. a `Series` or `Dataframe` with `dtype=StringDtype()` is now substantially fasterin my pandas fork than with object strings and pyarrow strings, since there's no overhead from the ExtensionArray interface. I've also implemented an `ExtensionArray` (mostly so I can use the extension tests in pandas to ferret out places in numpy that need to be fixed), and that has performance comparable to object strings and pyarrow strings.

String method operations (e.g. `series.str.capitalize()`) have performance comparable to object strings but have worse performance than pyarrow strings where a fast implementation in pyarrow is available. In principle numpy could match performance by implementing string operations as ufuncs, but for now I'm punting on that since it will require bikeshedding a public API and figuring out whether numpy should depend ICU, a lightweight unicode library, or should implement core unicode algorithms from scratch.

Memory usage is improved compared with object strings and pyarrow strings, and substantially improved compared with numpy fixed-width unicode strings, especially in worst-case situations like many short strings and one very long strings or mostly ASCII data. That's measured using `asv`'s peakmem benchmarks, which use the maximum resident size for the process as reported by the kernel, so it includes the binary size. Most of the memory usage in the pandas benchmarks is the binaries imported by pandas and the differences in memory usage in the benchmarks are much smaller than the total memory used by the process. Unfortunately more fine-grained benchmarking is very difficult as only some allocations in all these libraries participate in python's allocation tracking machinery.

Reviewer 4: Thanks for explaining

Algorithm-wise, how mature is it? Do you think it could be a drop-in alternative to `pyarrow[string]`, or is there anything not implemented / that would work differently / which hasn't been extensively tested?

Reviewer 3: I have a fork of pandas where I’ve added a `string[numpy]` ExtensionArray based on my dtype. It passes the full ExtensionArray test suite and almost passes all of the additional string array tests, the remaining failures are from a missing cast to/from datetimes (I’m working on support for that) and converting to/from PyArrow arrays, which I just haven’t looked at yet.

My hope is that it will be a drop-in replacement for `string[python]`, any ways in which it behaved differently I’d consider a bug. I’m not aware of any fundamental issues that would block making it a transparent drop-in replacement.

Right now I’m working on a NEP to propose adding the dtype to NumPy for NumPy 2.0 at the end of this year.

Reviewer 5: IMO the efficient algorithms that PyArrow brings are even more important for us than simply the ``string[python]`` behavior. How does the ExtensionArray on your fork compare to the ``string[pyarrow]`` implementation? Performance and memory usage would be interesting, although I am assuming that there shouldn't be a huge memory difference?

Reviewer 3: > How does the ExtensionArray on your fork compare to the string[pyarrow] implementation?

The memory usage should be comparable with pyarrow strings. Both are storing UTF8 bytestreams internally. I don't know offhand if arrow uses the small string optimization (storing the string content in the space normally reserved for a pointer to the string). It's difficult to compare memory usage exactly since the operating system facilities for this only allow you to measure the peak memory usage of a process and not all allocations necessarily use Python's allocation tracking machinery. I'm hoping to do a more careful memory usage benchmark as part of the NEP I'm writing.

The main difference in the storage is that right now I'm using individual heap allocations for each string array entry. Arrow just does a single allocation for all the array entries and has a secondary array of offsets to find the data for each string array element. I've thought a bit about following that approach, but it would mean we would have to either disallow mutating string arrays or there or have pathological behavior where enlarging a single array element could cause the entire array to get reallocated. It would also be nice to be able to use the short string optimization where a bitflag indicates whether an array element contains a pointer to heap data or contains a short string that is only a few bytes long, eliminating the need for a heap allocation for short strings.

For performance, do you mean for string manipulation operations like case folding or padding? In principle NumPy could add string ufuncs that would allow for fast implementations, but right now NumPy doesn't have a namespace for that. Currently, all the comparison operators are implemented as ufuncs, but no other string functionality is. There are string manipulation functions in the `np.char` namespace, but they just do a for loop over the array elements and call string functions on the scalars. 

I don't want to promise that string ufuncs definitely will happen in the future, but there's no real technical blockers, just social ones. NumPy doesn't currently have any ufuncs that only make sense with string data, so some thought needs to go into where in the namespace they should go. It will also require a decent amount of implementation work to add the functions, although mostly just tedious C coding.

Overall the goal is to facilitate a straightforward transition from workflows that used object string arrays while enabling possible performance improvements in the future that are currently impossible with object string arrays.

Reviewer 5: Generally, pandas is moving away from mutability in some sense (the CoW adoption), so that isn't very high on my priority list.

While a storage efficient string dtype is nice, this is kind of pointless if the operations aren't fast from a pandas PoV. One of the biggest advantages of arrow is that we can reduce memory but also that most operations are significantly faster, depending on what you are doing it's can be an order of magnitude. 

I am referring mostly to stuff like the ``str`` accessor but also things like factorization etc. 

So even if NumPy strings are ready in around a year (or some other time period), that's not helpful for us as long as NumPy does not ship fast algorithms on top of it. 

Sorry if this sounds harsh, that wasn't my intention. But having the string dtype without algorithms gets us only half the way compared to what PyArrow does, so this isn't a compelling argument to avoid making Arrow strings the default.

Reviewer 6: As a minimum, a fast regex engine could potentially help as some of the str accessor functions were (maybe still are) implemented using regex for string[pyarrow] where the functions did not exist in PyArrow (or the minimum version supported at the time).

Reviewer 5: Are you suggesting to implement this in pandas? That's something I personally don't have any interest in doing and would also be at least 0- on adding for the time being. Having this stuff in Arrow is nice since it reduces maintenance burden and also having better test coverage since more libraries will depend on it

Reviewer 6: I suspect a regex engine would be implemented in Numpy and then any str accessor functions not implemented in NumPy could be implemented using either regex or object fallback in pandas (just like we did for PyArrow initially).

# PDEP-10: PyArrow as a required dependency for default string inference implementation

- Created: 17 April 2023
- Status: Under discussion
- Discussion: [#52711](https://github.com/pandas-dev/pandas/pull/52711)
              [#52509](https://github.com/pandas-dev/pandas/issues/52509)
- Author: [Matthew Roeschke](https://github.com/mroeschke)
          [Patrick Hoefler](https://github.com/phofl)
- Revision: 1

## Abstract

This PDEP proposes that:

- PyArrow becomes a runtime dependency starting with pandas 3.0
- The minimum version of PyArrow supported starting with pandas 3.0 is version 7 of PyArrow.
- When the minimum version of PyArrow is bumped, PyArrow will be bumped to the highest version that has
  been released for at least 2 years.
- Starting in pandas 2.1, pandas raises a ``FutureWarning`` when needing to infer string data that the future

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 7: > don't we want to go further, and raise a `FutureWarning` upon `import pandas` if `pyarrow` isn't installed, warning that in the future it will become a required dependency?

I agree with Marco here.  I'd also suggest that if we go that route, the message points to a Github issue where we can gather feedback

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 5: Yeah a feedback issue is a very good idea.

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 2: I think feedback is also a great idea, but isn't raising a warning on import so soon after just releasing 2.0 for the next major release counterproductive for the whole user experience? Not aware of any other solution but I think this might cause a lot of frustrations.

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 8: I'm -1 on the warning on import pandas. I assume the majority of pandas users don't care if pyarrow is installed or not, and we'll be annoying them for no good reason.

For pip, I think we can start packaging arrow with the default pandas, and maybe have a pandas[noarrow] for users who want to avoid pyarrow until it's required. Not sure if for conda-forge it's so easy. If we do that, then the warning would make sense to me.

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 5: We could hold off on the import warning til 2.2 and only raise a warning whenever we would encounter a change in behavior (specifically the string inference). But I guess this would happen to everybody, so not really less noisy

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 5: Dask did something similar for Arrow, we pull it by default when dask is installed

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 7: > I'm -1 on the warning on import pandas. I assume the majority of pandas users don't care if pyarrow is installed or not, and we'll be annoying them for no good reason.

My proposal is to only warn users if they import pandas AND they don't have pyarrow installed.  Then the question is how they silence the warning:

1.  (Preferred) - Install pyarrow
2.  Maybe we allow them to set an environment variable to silence it

Ideally, they do (1) and if there are issues, we get a report about it. 

The problem we have is that we don't know how many people will really be affected by requiring pyarrow to be installed.  If it is a small number, then we are creating pain for a large segment of the community who have to silence the warning. But if it is anything over 10% (IMHO), we need to hear from those people, and a warning up top is the best way to get it.

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer: Is the status of pyarrow in pandas still experimental?
(Sorry, haven't kept up with this)

For me, warning that we are going to require pyarrow signals that you should be able to use the pyarrow-backed arrays everywhere, and there seems to be quite a bit of buggy/notimplemented things currently.
(at least going by https://github.com/pandas-dev/pandas/issues?q=is%3Aissue+is%3Aopen+label%3ABug+label%3AArrow)

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer: Personally, I think I am in favor of starting to warn in 2.2.

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 8: > My proposal is to only warn users if they import pandas AND they don't have pyarrow installed. 

That's clear, still -1. I think a huge amount of pandas users just install pandas with pip install pandas or same with conda. They won't get pyarrow until they explicitly try to use pyarrow strings to a pyarrow reader, which only advanced users will do. So, almost every regular pandas user will get an intimidating warning just by importing the module, feeling they did something wrong when they didn't. A very poor user experience in my opinion. And I don't see the advantage. With the exception of few users I assume most people will just be happy with whatever pip install pandas requires, and if we internally require numpy, pyarrow or something else, they don't care.

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 7: > So, almost every regular pandas user will get an intimidating warning just by importing the module, feeling they did something wrong when they didn't.

That depends on how we word the warning.  If we say something like "You better install pyarrow now or everything will break", that will scare them.  If we say something like "Starting with pandas 3.0, pyarrow will become a required installed dependency for pandas. Install it now to identify any potential issues and to remove this warning. Report issues to https://github.com/pandas-dev/pandas/issues/xxxxx"  I don't think the latter is intimidating.

Having said that, I think the specifics of when this warning will appear should be detailed as part of this PDEP.

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 4: To be fair, "just" warning when inferring string data will result in practically every user seeing the warning anyway, so maybe that's enough (if I understand correctly?)

So, if I install pandas 2.1.0 and don't have pyarrow installed, then `pd.Series(['foo'])` would raise a `FutureWarning` telling me that in the future the default will be a pyarrow string dtype, and that to opt-in to the new behaviour I need to install pyarrow and set `dtype='string[pyarrow]'`? Whereas if I did have pyarrow installed, then the warning would just say to set `dtype='string[pyarrow]'`?

Setting `dtype=` everywhere to silence the warning could be quite a lot of work, maybe there's a simpler way for users to opt-in to this?

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Reviewer 2: @MarcoGorelli this is a good point and compromise. Since one of the arguments in favour of this PDEP regards string treatment this seems like a good place to put the warning and relay the message about feedback in a GH issue.
It also allows a pandas global variable to suppress warnings such as this rather than have to rely on environment variable to suppress an 'on import' warning.

>   data type result will be `ArrowDtype` with `pyarrow.string` instead of object

Author: Right now I am leaning towards @MarcoGorelli's suggestion of only warning when pandas needs to perform string inference in 2.1.  I also think it's a good medium of warning that pyarrow will be required _and_ where it will make a difference in 3.0.

# PDEP-10: PyArrow as a required dependency for default string inference implementation

- Created: 17 April 2023
- Status: Under discussion
- Discussion: [#52711](https://github.com/pandas-dev/pandas/pull/52711)
              [#52509](https://github.com/pandas-dev/pandas/issues/52509)
- Author: [Matthew Roeschke](https://github.com/mroeschke)
          [Patrick Hoefler](https://github.com/phofl)
- Revision: 1

## Abstract

This PDEP proposes that:

- PyArrow becomes a required runtime dependency starting with pandas 3.0
- The minimum version of PyArrow supported starting with pandas 3.0 is version 7 of PyArrow.
- When the minimum version of PyArrow is bumped, PyArrow will be bumped to the highest version that has
  been released for at least 2 years.
- The pandas 2.1 release notes will have a big warning that PyArrow will become a required dependency starting
  with pandas 3.0.
- Starting in pandas 2.2, pandas raises a ``FutureWarning`` when PyArrow is not installed in the users
  environment when pandas is imported. This will ensure that only one warning is raised and users can

>   easily silence it if necessary.

Reviewer 7: I recall some discussion we had on this PDEP about having the warning point to a GitHub issue where we could collect feedback on this requirement.  If we agree on this concept, I think it should be mentioned here.

Reviewer 5: Added

Reviewer 7: Do you mind adding that the warning will point to the feedback issue?

Reviewer 9: Apologies for the naive question, but I thought the purpose of this PDEP was to collect feedback on pyarrow as a required dependency? I understand visibility may be more if there's a link to a new issue in a warning in a future version of pandas, but to me, doing it that way (to collect feedback all over again) just seems like we're going to be having this discussion again in 6 months and will end up kicking kicking the can down the road.

```

*str.startswith*

```python
In[3]: %timeit ser_object.str.startswith("a")
136 ms ± 300 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In[4]: %timeit ser_string.str.startswith("a")
11 ms ± 19.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Another advantage is I/O. PyArrow engines in pandas can provide a significant speedup. Currently, the data
are cast to NumPy dtypes, which requires roundtripping when converting back to PyArrow strings explicitly, which
hinders performance.

**Memory**

PyArrow backed strings use significantly less memory. Dask developers investigated this [here](https://www.coiled.io/blog/pyarrow-strings-in-dask-dataframes).

Short summary: PyArrow strings required 1/3 of the original memory.

## Drawbacks

Including PyArrow would naturally increase the installation size of pandas. For example, installing pandas and PyArrow
using pip from wheels, numpy and pandas requires about `70MB`, and including PyArrow requires around `120MB`. An increase
of installation size would have negative implication using pandas in space-constrained development or deployment environments

> such as AWS Lambda.

Reviewer 4: > and including PyArrow requires around `120MB`

maybe I was just misinterpreting this. Could you please rephrase to "and including PyArrow requires an additional `120MB`", to make it unambiguous that "120M" refers to just pyarrow, rather than "pandas + numpy + pyarrow"?

Additionally, if a user is installing pandas in an environment where wheels are not available through a `pip install` or `conda install`,
the user will need to also build Arrow C++ and related dependencies when installing from source. These environments include

- Alpine linux (commonly used as a base for Docker containers)
- WASM (pyodide and pyscript)
- Python development versions

Lastly, pandas development and releases will need to be mindful of PyArrow's development and release cadance. For example when
supporting a newly released Python version, pandas will also need to be mindful of PyArrow's wheel support for that Python version
before releasing a new pandas version.

### PDEP-1 History

- 17 April 2023: Initial version

> - 8 May 2023: Changed proposal to make pyarrow required in pandas 3.0 instead of 2.1

Reviewer 5: I don't have a real opinion on this, there were some requests that we should include it in this case

> - 8 May 2023: Changed proposal to make pyarrow required in pandas 3.0 instead of 2.1

Reviewer 6: just an observation. I'll be downvoting this proposal anyway so no real need to change this.

# PDEP-10: PyArrow as a required dependency

- Created: 17 April 2023
- Status: Under discussion
- Discussion: [#52711](https://github.com/pandas-dev/pandas/pull/52711)
              [#52509](https://github.com/pandas-dev/pandas/issues/52509)
- Author: [Matthew Roeschke](https://github.com/mroeschke)
- Revision: 1

## Abstract

This PDEP proposes that:

- PyArrow becomes a runtime dependency starting pandas 2.1
- The minimum version of PyArrow supported starting pandas 2.1 is version 7.
- The minimum version of PyArrow will be bumped every major pandas release to the highest

>   PyArrow version that has been released for at least 2 years.

Reviewer 6: using the major.minor.patch terminology, major could be 2-3 years (ignoring for now the proposal by some to make this more frequent) and minor is 6-9 months.

It is not clear here, is the minimum supported version kept for all minor releases in this proposal? 

Near the tail end of the minor release cycle, the minimum supported version of pyarrow could be 5 years old?

Author: > It is not clear here, is the minimum supported version kept for all minor releases in this proposal?

Correct

## Background

> PyArrow is an optional dependency of pandas that provides a wide range of supplimental feature to pandas:

Reviewer 2: ```suggestion
PyArrow is an optional dependency of pandas that provides a wide range of supplemental features to pandas:
```

accelerate PyArrow-backed data in pandas, notibly string and datetime types.

As of pandas 2.0, one can feasibly utilize PyArrow as an alternative data representation to NumPy with advantages such as:

1. Consistent ``NA`` support for all data types
2. Broader support of data types such as ``decimal``, ``date`` and nested types

## Motivation

While all the functionality described in the previous paragraph is currently optional, PyArrow has significant integration into many areas
of pandas. With our roadmap noting that pandas strives for better Apache Arrow interoperability [^1] and many projects [^2], within or beyond the Python ecosystem, adopting or interacting with the Arrow format, making PyArrow a required dependency provides an additional signal of confidence in the Arrow
ecosystem to pandas users.

Additionally, requiring PyArrow would simplify the related development within pandas and potentially improve NumPy functionality that would be better suited
by PyArrow including:

- Avoiding runtime checking if PyArrow is available to perform PyArrow object inference during constructor or indexing operations

> - Avoiding NumPy object data types more by default for analogous types that have native PyArrow support such as decimal, binary, and nested types

Reviewer 10: I might be too optimistic, but having pyarrow as a required dependency has the potential to make the c/cython-code for read_csv and read_json obsolete (if they are on par and similarly fast).

Reviewer 11: that would be a compile time dependency which we are not contemplating at the current time; possibly could propose in the future

## Drawbacks

Including PyArrow would naturally increase the installation size of pandas. For example, installing pandas and PyArrow using pip from wheels, numpy and pandas
are about `70MB`, and PyArrow is around `120MB`. An increase of installation size would have negative impliciation using pandas in space-constrained development
or deployment environments such as AWS Lambda.

Additionally, if a user is installing pandas in an environment where wheels are not available and needs to build from source, the user will need to build Arrow C++ and related dependencies. These environments include

- Alpine linux (commonly used as a base for Docker containers)
- WASM (pyodide and pyscript)
- Python development versions

Lastly, pandas development and releases will need to be mindful of PyArrow's development and release cadance. For example when supporting a newly released Python version, pandas will also need to be mindful of PyArrow's wheel support for that Python version before releasing a new pandas version.

### PDEP-1 History

- 17 April 2023: Initial version

[^1] <https://pandas.pydata.org/docs/development/roadmap.html#apache-arrow-interoperability>

> [^2] <https://arrow.apache.org/powered_by/>

Reviewer 2: Just noting that the supplied link for Apache Arrow contains documentation for `pyarrow` which lists pandas 1.0 or higher as an optional dependency. It is slightly confusing to have cyclical dependencies - cross product documentation might need to be improved?

Reviewer 12: > It is slightly confusing to have cyclical dependencies - cross product documentation might need to be improved?

I am not sure there is anything to improve? Pandas and pyarrow have already been optional dependencies of each other in both directions for quite a while (and even with this PDEP, pandas will stay an optional dependency of pyarrow).

# PDEP-10: PyArrow as a required dependency

- Created: 17 April 2023
- Status: Under discussion
- Discussion: [#52711](https://github.com/pandas-dev/pandas/pull/52711)
              [#52509](https://github.com/pandas-dev/pandas/issues/52509)
- Author: [Matthew Roeschke](https://github.com/mroeschke)
- Revision: 1

## Abstract

This PDEP proposes that:

> - PyArrow becomes a runtime dependency starting pandas 2.1

Reviewer 7: ```suggestion
- PyArrow becomes a runtime dependency starting with pandas 2.1
```

> - The minimum version of PyArrow supported starting pandas 2.1 is version 7.

Reviewer: Would this version be consistent across the entire pandas API?

e.g. If I wanted to bump the pyarrow version for just the CSV parser to something higher, would I be able to do it?

Author: The minimum version would be consistent across the library, but IMO that shouldn't stop development of features that exist in newer versions of pyarrow (we already do this with version checking or try/except)

> - The minimum version of PyArrow supported starting pandas 2.1 is version 7.

Reviewer 7: ```suggestion
- The minimum version of PyArrow supported starting with pandas 2.1 is version 7 of PyArrow.
```

- The minimum version of PyArrow will be bumped every major pandas release to the highest
  PyArrow version that has been released for at least 2 years, and the minimum PyArrow version will be
  maintained for every minor version in the major version series.

## Background

PyArrow is an optional dependency of pandas that provides a wide range of supplemental features to pandas:

- Since pandas 0.21.0, PyArrow provided I/O reading functionality for Parquet
- Since pandas 1.2.0, pandas integrated PyArrow into the `ExtensionArray` interface to provide an optional string data type backed by PyArrow
- Since pandas 1.4.0, PyArrow provided I/0 reading functionality for CSV
- Since pandas 1.5.0, pandas provided an `ArrowExtensionArray` and `ArrowDtype` to support all PyArrow data types within the `ExtensionArray` interface

> - Since pandas 2.0.0, All I/O readers have the option to return PyArrow-backed data types, and many methods now utilize PyArrow compute functions to

Reviewer 7: ```suggestion
- Since pandas 2.0.0, all I/O readers have the option to return PyArrow-backed data types, and many methods now utilize PyArrow compute functions to
```

accelerate PyArrow-backed data in pandas, notibly string and datetime types.

As of pandas 2.0, one can feasibly utilize PyArrow as an alternative data representation to NumPy with advantages such as:

1. Consistent ``NA`` support for all data types

> 2. Broader support of data types such as ``decimal``, ``date`` and nested types

Reviewer 7: I think that in https://github.com/pandas-dev/pandas/issues/52429#issuecomment-1509395487, @jbrockmendel showed an example where performance can be better with string types.  Ideally, we'd have more examples like that, so probably adding a third item here would be worthwhile.

Reviewer 5: https://phofl.github.io/pydata-berlin/pydata-berlin-2023/intro.html#70

and the following slides.

## Motivation

While all the functionality described in the previous paragraph is currently optional, PyArrow has significant integration into many areas
of pandas. With our roadmap noting that pandas strives for better Apache Arrow interoperability [^1] and many projects [^2], within or beyond the Python ecosystem, adopting or interacting with the Arrow format, making PyArrow a required dependency provides an additional signal of confidence in the Arrow
ecosystem to pandas users.

Additionally, requiring PyArrow would simplify the related development within pandas and potentially improve NumPy functionality that would be better suited
by PyArrow including:

> - Avoiding runtime checking if PyArrow is available to perform PyArrow object inference during constructor or indexing operations

Reviewer 13: Are there any small code samples we can add to drive this point home? I think still we would make a runtime determination whether to return a pyarrow or numpy-backed object even if both are installed, no?

> - Avoiding NumPy object data types more by default for analogous types that have native PyArrow support such as decimal, binary, and nested types

Reviewer 7: And string??

# PDEP-10: PyArrow as a required dependency

- Created: 17 April 2023
- Status: Under discussion
- Discussion: [#52711](https://github.com/pandas-dev/pandas/pull/52711)
              [#52509](https://github.com/pandas-dev/pandas/issues/52509)
- Author: [Matthew Roeschke](https://github.com/mroeschke)
- Revision: 1

## Abstract

This PDEP proposes that:

- PyArrow becomes a runtime dependency starting with pandas 2.1
- The minimum version of PyArrow supported starting with pandas 2.1 is version 7 of PyArrow.

> - The minimum version of PyArrow will be bumped every major pandas release to the highest

Reviewer: Hmm.. This might be too aggressive and might also make it hard to predict what the minimum version will be.

I'd recommend following what we do for numpy, which is according to NEP 29, support
"all minor versions of NumPy released in the prior 24 months from the anticipated release date with a minimum of 3 minor versions of NumPy", for arrow as well.

Reviewer 13: I think the challenge with offering a similar support window for the two libraries is that NumPy has a very stable ABI whereas PyArrow does not

Reviewer 8: Not sure if I'm missing something, but sounds like what @lithomas1 is proposing is pretty much the same as what's written in the proposal but phrased in a different way. Has the proposal been updated, or am I misunderstanding that supporting the releases of the last 24 months, and supporting the highest/oldest version two years old?

Reviewer: I read this as every major release we will bump the min required version of pyarrow to the latest version, but might be misreading here.

Note that my proposed change would be different in that we would drop Arrow versions in both major/minor versions (as opposed to every major version), just like we do with numpy (once we reach the end of the NEP support window). 

> I think the challenge with offering a similar support window for the two libraries is that NumPy has a very stable ABI whereas PyArrow does not

I might have missed some more discussion on this, but I thought we were going to restrict current usage of pyarrow to just what's exposed through Python.

Reviewer 8: > I read this as every major release we will bump the min required version of pyarrow to the latest version, but might be misreading here.

To the latest version `that has been released for at least 2 years`. So, the minimum PyArrow version we support will be around 24 months old, and we should be supporting all the versions since that one, so more or less the same policy as NumPy. @mroeschke not sure if it's easy to rephrase in a way that it's more obvious what's the policy.

About bumping in major or minor releases, I don't have a preference, either is fine for me.

Reviewer: Can you consider this to upgrading pyarrow in both major and minor versions to be consistent with numpy?

I ask because it is probably tricky for downstream to predict the length of our major release cycle (for 2.0 I think we delayed it twice. IIRC 1.4 was supposed to be 2.0).

Author: > Can you consider this to upgrading pyarrow in both major and minor versions to be consistent with numpy?

Sure that would be okay with me too

PyArrow is an optional dependency of pandas that provides a wide range of supplemental features to pandas:

- Since pandas 0.21.0, PyArrow provided I/O reading functionality for Parquet
- Since pandas 1.2.0, pandas integrated PyArrow into the `ExtensionArray` interface to provide an optional string data type backed by PyArrow
- Since pandas 1.4.0, PyArrow provided I/0 reading functionality for CSV
- Since pandas 1.5.0, pandas provided an `ArrowExtensionArray` and `ArrowDtype` to support all PyArrow data types within the `ExtensionArray` interface
- Since pandas 2.0.0, all I/O readers have the option to return PyArrow-backed data types, and many methods now utilize PyArrow compute functions to
accelerate PyArrow-backed data in pandas, notibly string and datetime types.

As of pandas 2.0, one can feasibly utilize PyArrow as an alternative data representation to NumPy with advantages such as:

1. Consistent ``NA`` support for all data types
2. Broader support of data types such as ``decimal``, ``date`` and nested types

## Motivation

While all the functionality described in the previous paragraph is currently optional, PyArrow has significant integration into many areas

> of pandas. With our roadmap noting that pandas strives for better Apache Arrow interoperability [^1] and many projects [^2], within or beyond the Python ecosystem, adopting or interacting with the Arrow format, making PyArrow a required dependency provides an additional signal of confidence in the Arrow

Reviewer 8: Small thing, but it usually makes things easier (like visualizing the file in github or an IDE) if we limit the line length to < 99 or so, as we do with Python code

ecosystem to pandas users.

Additionally, requiring PyArrow would simplify the related development within pandas and potentially improve NumPy functionality that would be better suited
by PyArrow including:

- Avoiding runtime checking if PyArrow is available to perform PyArrow object inference during constructor or indexing operations
  - Currently, there are 17 `import_optional_dependency("pyarrow")` checks throughout the pandas code base
- Avoiding NumPy object data types more by default for analogous types that have native PyArrow support such as string, decimal, binary, and nested types

## Drawbacks

Including PyArrow would naturally increase the installation size of pandas. For example, installing pandas and PyArrow using pip from wheels, numpy and pandas
are about `70MB`, and PyArrow is around `120MB`. An increase of installation size would have negative impliciation using pandas in space-constrained development
or deployment environments such as AWS Lambda.

> Additionally, if a user is installing pandas in an environment where wheels are not available and needs to build from source, the user will need to build Arrow C++ and related dependencies. These environments include

Reviewer 12: I think it would be good to more explicitly say that you need to do this (installing Arrow C++) manually and this is possible through `pip install pyarrow` 
(there are other python packages that also have C/C++ code but that do that build automatically (if you have the dependencies such as a compiler) when installing from source)

Reviewer 4: thanks for providing details here - is this (`pip install pyarrow`) much of a hurdle for these cases where pandas wheels aren't available?

Reviewer 12: Whoops, sorry there was a "not" missing in "_not_ possible through `pip install pyarrow`. Corrected now. 

But on the actual question how much of a hurdle this is: I would say, try it out yourself :) That's the best way to get an idea of how difficult it is, otherwise you can only take (or not) my words in saying that: yes, this is a huge hurdle. Installing pyarrow and Arrow C++ from source is far from trivial.

Reviewer 4: thanks - from the installation instructions, it certainly seems tricky https://arrow.apache.org/docs/developers/cpp/building.html#

  optional string data type backed by PyArrow
- Since pandas 1.4.0, PyArrow provided I/0 reading functionality for CSV
- Since pandas 1.5.0, pandas provided an `ArrowExtensionArray` and `ArrowDtype` to support all PyArrow
  data types within the `ExtensionArray` interface
- Since pandas 2.0.0, all I/O readers have the option to return PyArrow-backed data types, and many methods
  now utilize PyArrow compute functions to
accelerate PyArrow-backed data in pandas, notibly string and datetime types.

As of pandas 2.0, one can feasibly utilize PyArrow as an alternative data representation to NumPy with advantages such as:

1. Consistent `NA` support for all data types
2. Broader support of data types such as `decimal`, `date` and nested types

Additionally, when users pass string data into pandas constructors without specifying a data type, the result data type
is `object`. With pyarrow string support available since 1.2.0, requiring pyarrow for 3.0 will allow pandas to default
the inferred type to the more efficient pyarrow string type.

```python
In [1]: import pandas as pd

In [2]: pd.Series(["a"]).dtype
Out[2]: dtype('O')

> ```

Reviewer 7: The context of this example is a little confusing.  I think you want to indicate that this is current behavior, and then provide an example of what the proposed new behavior would be as of version 3.0.

As of pandas 2.0, one can feasibly utilize PyArrow as an alternative data representation to NumPy with advantages such as:

1. Consistent `NA` support for all data types;
2. Broader support of data types such as `decimal`, `date` and nested types;
3. Better interoperability with other dataframe libraries based on Arrow.

## Motivation

While all the functionality described in the previous paragraph is currently optional, PyArrow has significant
integration into many areas of pandas. With our roadmap noting that pandas strives for better Apache Arrow
interoperability [^1] and many projects [^2], within or beyond the Python ecosystem, adopting or interacting with
the Arrow format, making PyArrow a required dependency provides an additional signal of confidence in the Arrow
ecosystem (as well as improving interoperability with it).

### Immediate User Benefit 1: pyarrow strings

Currently, when users pass string data into pandas constructors without specifying a data type, the resulting data type

> is `object`, which has horrendous memory and performance implications.

Reviewer 7: "horrendous" is a bit of a strong word.  I think it would be better to say "which has poor memory usage and performance as compared to pyarrow strings."

Reviewer 5: IMO horrendous is appropriate. I observed up to 5x more memory consumption with NumPy object and performance is just terrible if you compare head-to-head.

I did a bunch of benchmarks here: https://medium.com/towards-data-science/utilizing-pyarrow-to-improve-pandas-and-dask-workflows-2891d3d96d2b

Just one example: Avoiding the Arrow strings -> object conversion in read_parquet reduces the runtime by 2/3 ( I am talking about the runtime of the whole ``read_parquet`` call, e.g. the conversion actually takes __at least twice as long__ as I/O.

"poor" isn't strong enough imo 😅

Reviewer 7: > "poor" isn't strong enough imo 😅

Fair enough, but it just didn't feel right to have the word "horrendous" in a PDEP.  The reality is that we have been living with this bad memory and performance for years, so we only know it is horrendous because of the comparison.   So maybe if you said "which has significantly much worse memory usage and performance as compared to pyarrow strings." it would feel right.

Author: +1 to `"which has significantly much worse memory usage and performance as compared to pyarrow strings."`

Reviewer 5: Updated with Matts suggestion
